[FLINK-9384]KafkaAvroTableSource failed to work due to type mismatch

tragicjun opened the pull request:
## What is the purpose of the change
An exception was thrown when using KafkaAvroTableSource, complaining like:
```
 "TableSource of type org.apache.flink.streaming.connectors.kafka.Kafka011AvroTableSource returned a DataStream of type GenericType<org.apache.flink.types.Row> that does not match with the type Row(id: Integer, name: String, age: Integer, event: GenericType<java.util.Map>) declared by the TableSource.getReturnType() method. Please validate the implementation of the TableSource."
```
This pull request fixes the issue by referencing the implementation of KafkaJsonTableSource. 

## Brief change log

  - Add a TypeInformation as the second argument to the constructor of AvroRowDeserializationSchema
  - AvroRowDeserializationSchema overrides the method getProducedType()
  - KafkaAvroTableSource constructs a TypeInformation converted from AvroRecordClass and pass it to AvroRowDeserializationSchema

## Verifying this change

This change is already covered by existing tests, such as KafkaAvroTableSourceTestBase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

suez1224 commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaAvroTableSource.java:
```diff
@@ -81,7 +85,12 @@ public String explainSource() {
+	}
+
+	/** Converts the table schema into into the return type. */
+	private static RowTypeInfo tableSchemaToReturnType(TableSchema tableSchema) {
```
Do you need this extra function, I think just inlining it should be fine.

suez1224 commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java:
```diff
@@ -110,6 +116,8 @@ private void writeObject(ObjectOutputStream oos) throws IOException {
 		oos.writeObject(recordClazz);
 	}
 
+
```
extra blank line here.

tragicjun commented:
Thanks @suez1224 for reviewing. I've committed a new patch as per your suggestions. As for the unit test, the issue was exposed in my integration tests which requires a kafka setup, any advice on how it could be done in unit tests?

tragicjun commented:
Hi @suez1224, I've added some checks in the unit tests(KafkaAvroTableSourceTestBase), which would fail before the fix. Meanwhile, I found an issue in the previous version (specifically,  the produced type info should be inferred from AvroRecordClass not TableSchema), which has been fixed. Would you please take a review?

twalthr commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java:
```diff
@@ -76,19 +77,35 @@
+	/** Type information describing the result type.
+	 *
+	 */
+	private final TypeInformation<Row> typeInfo;
```
Declare this transient?

twalthr commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java:
```diff
@@ -76,19 +77,35 @@
 	 */
 	private SpecificRecord record;
 
+	/** Type information describing the result type.
```
Put the text one line down.

twalthr commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java:
```diff
@@ -76,19 +77,35 @@
+	 * @param recordClazz Avro record class used to deserialize Avro's record to Flink's row
+	 * @param typeInfo Type information describing the result type.
+	 */
+	public AvroRowDeserializationSchema(Class<? extends SpecificRecord> recordClazz, TypeInformation<Row> typeInfo) {
```
This constructor is not required. We should always use the result of the `AvroRecordClassConverter`.

tragicjun commented:
@twalthr Thanks for the comment. I've commited a new version as per your suggestion. Particularly, I agree that TypeInfomation should always be derived from the 'avroRecrodClass', but the constructor of AvroRowDeserializationSchema only takes argument as type 'Class<? extends SpecificRecord>'. In this version I've changed the argument as 'Class<? extends SpecificRecordBase>' and fixed relevant test cases accordingly. This version of fix has passed relevant unit tests and my local integration tests.

twalthr commented:
Thank you @tragicjun. LGTM. Merging...

tragicjun commented:
@twalthr sure, just a second, I never did this before, let me try

tragicjun commented:
@twalthr the latest commit could be merged which contains all the changes.
